### PR TITLE
[core] Retry failed redis request

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

After https://github.com/ray-project/ray/pull/35123 the request in ray is serialized and we should be able to retry the failed redis request in ray.

This PR refactor the RedisContext a little bit by remove the CallbackItem and introduce the RequestContext. In side the RequestContext, the failed request will be retried automatically.

If still failed in the end, it'll just crash.

## Related issue number
https://github.com/ray-project/ray/issues/34014

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -311,6 +311,12 @@ RAY_CONFIG(int, worker_niceness, 15)
 RAY_CONFIG(int64_t, redis_db_connect_retries, 600)
 RAY_CONFIG(int64_t, redis_db_connect_wait_milliseconds, 100)
 
+/// Number of retries for a redis request failure.
+RAY_CONFIG(int64_t, num_redis_request_retries, 3)
+
+/// Interval between two retries.
```
Comment (scv119) [inline src/ray/common/ray_config_def.h:317]: would 300ms retry catch the normal failure cases? i'd expect exponential back-offs with longer delays to work.

File: src/ray/gcs/redis_context.cc [hunk 2]
```diff
@@ -172,55 +152,66 @@ const std::vector<std::optional<std::string>> &CallbackReply::ReadAsStringArray(
...
+    RAY_LOG(FATAL) << "Failed to run redis cmds: [" << absl::StrJoin(redis_cmds_, " ")
+                   << "] for " << RayConfig::instance().num_redis_request_retries()
+                   << " times.";
+  }
+
+  --pending_retries_;
+
+  auto fn =
+      +[](struct redisAsyncContext *async_context, void *redis_reply, void *privdata) {
+        auto *request_cxt = (RedisRequestContext *)privdata;
```
Comment (scv119) [inline src/ray/gcs/redis_context.cc:182]: use static_cast and handle errors?

nit: request_ctx

File: src/ray/gcs/redis_context.cc [hunk 3]
```diff
@@ -550,26 +541,14 @@ std::unique_ptr<CallbackReply> RedisContext::RunArgvSync(
...
-      RedisCallbackManager::instance().AddCallback(redis_callback, io_service_);
-  // Run the Redis command.
-  Status status = redis_async_context_->RedisAsyncCommandArgv(
-      reinterpret_cast<redisCallbackFn *>(&GlobalRedisCallback),
-      reinterpret_cast<void *>(callback_index),
-      args.size(),
-      argv.data(),
-      argc.data());
-  return status;
+  auto request_context = new RedisRequestContext(io_service_,
```
Comment (scv119) [inline src/ray/gcs/redis_context.cc:547]: can we create a shared_ptr for request_context and use std::enable_shared_from_this for RedisRequestContext?

File: src/ray/gcs/redis_context.cc (hunk 2 above)
Comment (iycheng) [reply src/ray/gcs/redis_context.cc:182]: static cast doesn't return anything error I think?
Here we need `reinterpreter_cast` which is the same as this type conversion. (I can update it to reinterpreter_cast)

The issue here is that it should be accepted as C callback.

I guess talking with C library, you basically lose all smart pointer features.

File: src/ray/gcs/store_client/test/redis_store_client_test.cc [hunk 4]
```diff
@@ -55,153 +87,6 @@ TEST_F(RedisStoreClientTest, AsyncGetAllAndBatchDeleteTest) {
   TestAsyncGetAllAndBatchDelete();
 }
 
-TEST_F(RedisStoreClientTest, BasicSimple) {
```
Comment (iycheng) [inline src/ray/gcs/store_client/test/redis_store_client_test.cc:58]: merge issues. I'll revert this part back.

**Comment by iycheng** [comment]
Retry failed request seems fine but in the test added it's about disconnect and it's not trivial to support reconnect in some way.

https://github.com/redis/hiredis/blob/master/README.md#connecting-1

I'll see how to update the test so redis reject but not disconnect.

File: src/ray/common/ray_config_def.h (hunk 1 above)
Comment (iycheng) [reply src/ray/common/ray_config_def.h:317]: I can update it to that. I was thinking that for the short term support of this, we only cover the short duration failure.
Here reconnection won't be supported, so only certain network failure can be supported.

File: src/ray/gcs/redis_context.cc [hunk 5]
```diff
@@ -172,55 +152,71 @@ const std::vector<std::optional<std::string>> &CallbackReply::ReadAsStringArray(
...
+RedisRequestContext::RedisRequestContext(instrumented_io_context &io_service,
+                                         RedisCallback callback,
+                                         RedisAsyncContext *context,
+                                         std::vector<std::string> args)
+    : exp_back_off_(RayConfig::instance().redis_retry_base_ms(),
+                    RayConfig::instance().redis_retry_multiplier(),
+                    RayConfig::instance().redis_retry_max_ms()),
+      io_service_(io_service),
+      redis_context_(context),
+      pending_retries_(RayConfig::instance().num_redis_request_retries()),
```
Comment (scv119) [inline src/ray/gcs/redis_context.cc:164]: should we always initialize pending_retries_ to at least 1?

**Review by scv119 — APPROVED** [review]

File: src/ray/gcs/redis_context.cc (hunk 5 above)
Comment (iycheng) [reply src/ray/gcs/redis_context.cc:164]: yea, i think we should defined the num_redis_request_retries to be the retris we'll do if the first failed.

**Pull request merged**